packages/headless/src/controllers/core/generated-answer/headless-core-generated-answer.ts

```diff
@@ -316,7 +324,20 @@ export function buildCoreGeneratedAnswer(
     },
 
     sendFeedback(feedback) {
-      dispatch(analyticsClient.logGeneratedAnswerFeedback(feedback));
+      if ((feedback as GeneratedAnswerFeedbackV2).helpful !== undefined) {
```

Comment:
Do we really need two actions?

I think we could always dispatch the same action, but the `feedback` argument would be `GeneratedAnswerFeedback | GeneratedAnswerFeedbackV2`. Then in the action's body, you'd make / submit the proper action based on the received `feedback` type.

E.g.,

```typescript
function isGeneratedAnswerFeedbackV2(
  feedback: GeneratedAnswerFeedback | GeneratedAnswerFeedbackV2
): feedback is GeneratedAnswerFeedbackV2 {
  return typeof feedback === 'object';
}

export const logGeneratedAnswerFeedback = (
  feedback: GeneratedAnswerFeedback | GeneratedAnswerFeedbackV2
): CustomAction =>
  makeAnalyticsAction({
    prefix: 'analytics/generatedAnswer/sendFeedback',
    __legacy__getBuilder: (client, state) => {
      const generativeQuestionAnsweringId =
        generativeQuestionAnsweringIdSelector(state);
      if (!generativeQuestionAnsweringId) {
        return null;
      }
      return isGeneratedAnswerFeedbackV2(feedback)
        ? client.makeGeneratedAnswerFeedbackSubmitV2({
            generativeQuestionAnsweringId,
            ...feedback,
          })
        : client.makeGeneratedAnswerFeedbackSubmit({
            generativeQuestionAnsweringId,
            reason: feedback,
          });
    },
  });
```

This way in V3, instead of removing an analytics action, you'd be restricting the type this action accepts as an argument.

In your tests, you'd just be checking that the action gets dispatched with the expected payload.

Reply:
indeed much better! thank you